File: src/runtime/python/input_buffer.py
#! coding=UTF-8

# Copyright 2012 Thomas Bollmeier <tbollmeier@web.de>

# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at

#   http://www.apache.org/licenses/LICENSE-2.0

# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from .util import DynamicObject
            
class InputBuffer(object):

    def __init__(self, stream, fillSize = 1):
         
        self._stream = stream
        self._fillSize = fillSize
        
        self._content = ""
        self._line = 1
        self._column = 0
            
    def setFillSize(self, newFillSize):
        
        self._fillSize = newFillSize
        
    def getContent(self):

        self._fillContent()
        
        return self._content[0:self._fillSize]

    def getPositionInfo(self):
        
        res = DynamicObject()
        res.line = self._line
        res.column = self._column
        
        return res
    
    def consumeChar(self):
        
        if self._content:
            
            ch = self._content[0]
            
            if len(self._content) > 1:
                self._content = self._content[1:]
            else:
                self._content = ""
                
            self._fillContent()
            
            self._updatePosInfo(ch)
                
            return ch
        
        else:
            
            return ""
        
    def consumeAll(self):
        
        if len(self._content) <= self._fillSize:
            
            res = self._content
            self._content = ""
            
        else:
            
            res = self._content[0:self._fillSize]
            self._content = self._content[self._fillSize:]
        
        for ch in res:
            self._updatePosInfo(ch)
                
        return res

    def _fillContent(self):

        while len(self._content) < self._fillSize:
            if self._stream.endOfInput():
                break
            self._content += self._stream.getNextChar()
            
    def _updatePosInfo(self, ch):
        
        if ch != "\n":
            self._column += 1
        else:
            self._line += 1
            self._column = 0

File: src/runtime/php/input_buffer.php
<?php

/*
 * Copyright 2012 Thomas Bollmeier <tbollmeier@web.de>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

class Bovinus_InputBuffer {

    public function __construct($stream, $fillSize = 1) {

        $this->_stream = $stream;
        $this->_fillSize = $fillSize;
        $this->_positionInfo = new Bovinus_PositionInfo(1, 0);
    }

    public function setFillSize($fillSize) {

        $this->_fillSize = $fillSize;
    }

    public function getContent() {

        $this->fillContent();

        return substr($this->_content, 0, $this->_fillSize);
    }

    public function getPositionInfo() {

        return $this->_positionInfo;
    }

    public function consumeChar() {

        $ch = '';

        $len = strlen($this->_content);

        if ($len > 0) {
            $ch = $this->_content[0];
            if ($len > 1) {
                $this->_content = substr($this->_content, 1);
            } else {
                $this->_content = "";
            }
        }

        $this->fillContent();

        if ($ch !== "\n") {
            $this->_positionInfo->column++;
        } else {
            $this->_positionInfo->line++;
            $this->_positionInfo->column = 0;
        }

        return $ch;
    }

    public function consumeNChars($n) {

        $res = "";

        for ($i = 0; $i < $n; $i++) {
            $res .= $this->consumeChar();
        }

        return $res;
    }

    public function consumeAll() {

        $res = "";

        if (strlen($this->_content) <= $this->_fillSize) {
            $res = $this->_content;
            $this->_content = "";
        } else {
            $res = substr($this->_content, 0, $this->_fillSize);
            $this->_content = substr($this->_content, $this->_fillSize);
        }

        for ($i = 0; $i < strlen($res); $i++) {
            $ch = $res[$i];
            if ($ch !== "\n") {
                $this->_positionInfo->column++;
            } else {
                $this->_positionInfo->line++;
                $this->_positionInfo->column = 0;
            }
        }

        return $res;
    }

    private function fillContent() {

        while (strlen($this->_content) < $this->_fillSize) {
            if ($this->_stream->endOfInput())
                break;
            $this->_content .= $this->_stream->getNextChar();
        }
    }

    private $_stream;
    private $_fillSize;
    private $_content = "";
    private $_positionInfo;

}

class Bovinus_PositionInfo {

    public $line;
    public $column;

    public function __construct($line, $column) {

        $this->line = $line;
        $this->column = $column;
    }

}
